cmd/microcloud/status.go

```diff
@@ -0,0 +1,444 @@
+// Finally, the status will be Success, implying no warnings.
+func (w Warnings) Status() StatusLevel {
+	if len(w) == 0 {
+		return Success
+	}
+
+	for _, warning := range w {
+		if warning.Level == Error {
+			return Error
+		}
+	}
+
+	return Warn
+}
+
+// StatusLevel represents the severity level of warnings.
+type StatusLevel int
+
+const (
+	// Success represents a lack of warnings.
+	Success StatusLevel = iota
+
+	// Warn represents a medium severity warning.
+	Warn
+
+	// Error represents a critical warning.
+	Error
+)
+
+// Symbol returns the single-character symbol representing the StatusLevel, color coded.
+func (s StatusLevel) Symbol() string {
+	switch s {
+	case Success:
+		return style.SuccessSymbol()
+	case Warn:
+		return style.WarningSymbol()
+	case Error:
+		return style.ErrorSymbol()
+	}
+
+	return ""
+}
+
+// Symbol returns a word representing the StatusLevel, color coded.
+func (s StatusLevel) String() string {
+	switch s {
+	case Success:
+		return style.SuccessColor("HEALTHY")
+	case Warn:
+		return style.WarningColor("WARNING")
+	case Error:
+		return style.ErrorColor("ERROR")
+	}
+
+	return ""
+}
+
+type cmdStatus struct {
+	common *CmdControl
+}
+
+func (c *cmdStatus) Command() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "status",
+		Short: "Deployment status with configuration warnings",
+		RunE:  c.Run,
+	}
+
+	return cmd
+}
+
+func (c *cmdStatus) Run(cmd *cobra.Command, args []string) error {
+	if len(args) != 0 {
+		return cmd.Help()
+	}
+
+	cloudApp, err := microcluster.App(microcluster.Args{StateDir: c.common.FlagMicroCloudDir})
+	if err != nil {
+		return err
+	}
+
+	status, err := cloudApp.Status(context.Background())
+	if err != nil {
+		return fmt.Errorf("Failed to get MicroCloud status: %w", err)
+	}
+
+	if !status.Ready {
+		return fmt.Errorf("MicroCloud is uninitialized, run 'microcloud init' first")
+	}
+
+	cfg := initConfig{
+		autoSetup: true,
+		bootstrap: false,
+		common:    c.common,
+		asker:     &c.common.asker,
+		systems:   map[string]InitSystem{},
+		state:     map[string]service.SystemInformation{},
+	}
+
+	cfg.name = status.Name
+	cfg.address = status.Address.Addr().String()
+
+	services := []types.ServiceType{types.MicroCloud, types.LXD}
+	optionalServices := map[types.ServiceType]string{
+		types.MicroCeph: api.MicroCephDir,
+		types.MicroOVN:  api.MicroOVNDir,
+	}
+
+	services, err = cfg.askMissingServices(services, optionalServices)
+	if err != nil {
+		return err
+	}
+
+	// Instantiate a handler for the services.
+	sh, err := service.NewHandler(status.Name, status.Address.Addr().String(), c.common.FlagMicroCloudDir, c.common.FlagLogDebug, c.common.FlagLogVerbose, services...)
+	if err != nil {
+		return err
+	}
+
+	cloudClient, err := sh.Services[types.MicroCloud].(*service.CloudService).Client()
+	if err != nil {
+		return err
+	}
+
+	// Query the status API for the cluster.
+	statuses, err := client.GetStatus(context.Background(), cloudClient)
```

Comment:
I have deployed a MicroCloud with three members but shutdown one of them. 
Now when I run `microcloud cluster list` I see one of them being `UNREACHABLE` whilst `microcloud status` gives me `Error: Get "https://10.134.197.4:9443/1.0/status": Unable to connect to "10.134.197.4:9443": dial tcp 10.134.197.4:9443: connect: no route to host`. 

I would have expected to still see the status table but with one member giving the status `OFFLINE`.
It's my assumption that the error is raised at this point :)

Reply:
After starting up the member I shortly saw this:

```
root@m1:/var/snap/microcloud/common/state# microcloud status

 Status: WARNING

 ┃ ! m3 is not available

┌──────┬────────────────┬───────────┬────────────────────┬────────────────────────┬─────────┐
│ Name │ Address        │ OSD Disks │ MicroCeph Services │ MicroOVN Services      │ Status  │
├──────┼────────────────┼───────────┼────────────────────┼────────────────────────┼─────────┤
│ m2   │ 10.134.197.5   │ 1         │ mds,mgr,mon        │ central,chassis,switch │ ONLINE  │
│ m3   │ 10.134.197.4   │ 1         │ mds,mgr,mon        │ central,chassis,switch │ Offline │
│ m1   │ 10.134.197.164 │ 1         │ mds,mgr,mon        │ central,chassis,switch │ ONLINE  │
└──────┴────────────────┴───────────┴────────────────────┴────────────────────────┴─────────┘
```

So I guess there is a check for this but it looks to be failing currently if the entire node is offline.

Reply:
Yeah this was a bug. If a cluster member can't be reached, we now populate it with empty node-specific information instead, and list it as unreachable.

Reply:
Great thanks, I'll give it another shot :)